File: programs/list/src/instructions/push_element.rs
use {
    anchor_lang::{
        prelude::*,
        solana_program::system_program
    },
    crate::state::*,
    std::mem::size_of
};

#[derive(Accounts)]
#[instruction(
    value: Pubkey,
    bump: u8,
)]
pub struct PushElement<'info> {
    #[account(
        mut, 
        seeds = [
            SEED_LIST,
            list.owner.key().as_ref(), 
            list.namespace.as_ref()
        ],
        bump = list.bump, 
        has_one = owner,
    )]
    pub list: Account<'info, List>,

    #[account(mut)]
    pub owner: Signer<'info>,

    #[account(mut)]
    pub payer: Signer<'info>,

    #[account(
        init,
        seeds = [
            SEED_ELEMENT,
            list.key().as_ref(),
            list.count.to_be_bytes().as_ref(),
        ],
        bump = bump,
        payer = payer,
        space = 8 + size_of::<Element>(),
    )]
    pub element: Account<'info, Element>,
    
    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<PushElement>,
    value: Pubkey,
    bump: u8,
) -> ProgramResult {
    // Get accounts.
    let element = &mut ctx.accounts.element;
    let list = &mut ctx.accounts.list;
    
    // Initialize element account.
    element.index = list.count;
    element.value = value;
    element.bump = bump;

    // Increment list counter.
    list.count = list.count.checked_add(1).unwrap();
    
    return Ok(());
}

File: programs/indexer/client/src/instruction/pushElement.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { Indexer } from "../idl";
import { Account } from "../account";

export type PushElementArgs = {
  list: PublicKey;
  owner: PublicKey;
  value: PublicKey;
};

export class PushElement {
  private account: Account;
  private program: Program<Indexer>;

  constructor(account: Account, program: Program<Indexer>) {
    this.account = account;
    this.program = program;
  }

  public async pushElement({
    list,
    owner,
    value,
  }: PushElementArgs): Promise<TransactionInstruction> {
    const listData = await this.account.list.data(list);
    const elementPDA = await this.account.element.pda(list, listData.count);
    return this.program.instruction.pushElement(value, elementPDA.bump, {
      accounts: {
        list: list,
        element: elementPDA.address,
        owner: owner,
        payer: owner,
        systemProgram: SystemProgram.programId,
      },
    });
  }
}
